azurelinuxagent/ga/exthandlers.py

```diff
@@ -365,31 +365,39 @@ def handle_ext_handlers(self, etag=None):
 
         self.ext_handlers.extHandlers.sort(key=operator.methodcaller('sort_key'))
         for ext_handler in self.ext_handlers.extHandlers:
+            # Use updated time limit for each dependency level
             if dep_level != ext_handler.sort_key():
                 dep_level = ext_handler.sort_key()
                 deps_wait_until = handlers_wait_until
-                handlers_wait_until += datetime.timedelta(seconds=(DEFAULT_EXT_TIMEOUT_MINUTES * 60))
+                handlers_wait_until += datetime.timedelta(minutes=DEFAULT_EXT_TIMEOUT_MINUTES)
+
+            # Make sure the previous level extensions are installed (or timed out)
             self.wait_on_ext_handler_dependencies(ext_handler, deps_wait_until)
+
             self.handle_ext_handler(ext_handler, etag)
 
     def wait_on_ext_handler_dependencies(self, ext_handler, wait_until):
+        # Collect all the dependencies in the handler
+        # Check the status of the extension in each dependency
+        # Wait until it becomes success or times out
+
         dependencies = sum([e.dependencies for e in ext_handler.properties.extensions], [])
         for dependency in dependencies:
             handler_i = ExtHandlerInstance(dependency.handler, self.protocol)
-            if dependency.timeout is not None and dependency.timeout < DEFAULT_EXT_TIMEOUT_MINUTES:
-                wait_until -= datetime.timedelta(seconds=( (DEFAULT_EXT_TIMEOUT_MINUTES - dependency.timeout) * 60))
-            for ext in dependency.exts:
-                success_status, status = handler_i.is_ext_status_success(ext)
-                while not success_status:
-                    if datetime.datetime.utcnow() > wait_until:
-                        raise ExtensionError("Timeout waiting for success status "
-                                             "from dependency {}/{} for {}"
-                                             "status was: {}".format(
-                                                 dependency.handler.name, ext.name,
-                                                 ext_handler.name, status))
-                    else:
-                        time.sleep(10)
-                    success_status, status = handler_i.is_ext_status_success(ext)
+            success_status, status = handler_i.is_ext_status_success(dependency.ext)
+
+            # Keep polling for the extension status until it becomes success or times out
+            while not success_status and datetime.datetime.utcnow() <= wait_until:
+                time.sleep(5)
+                success_status, status = handler_i.is_ext_status_success(dependency.ext)
+
+            # In case of timeout, we log it and continue with the rest of the extensions
```

Comment:
This is not the correct behaviour.

The extension should *not* be run if its dependency timed out. The extension runs once this function returns, unless this function throws an exception.